api/v1alpha1/directivebreakdown_types.go

```diff
@@ -90,39 +80,72 @@ type AllocationSetComponents struct {
 	Constraints AllocationSetConstraints `json:"constraints,omitempty"`
 }
 
-// DirectiveBreakdownSpec defines the storage information WLM needs to select NNF Nodes and request storage from the selected nodes
-type DirectiveBreakdownSpec struct {
-	// DW is the Datawarp Directive Record
-	DW DWRecord `json:"dwRecord"`
-
-	// Name is the identifier for this directive breakdown
-	Name string `json:"name"`
-
-	// Type is the type specified in the #DW directive
-	// +kubebuilder:validation:Enum=raw;xfs;gfs2;lustre
-	Type string `json:"type"`
+const (
+	StorageLifetimePersistent = "persistent"
+	StorageLifetimeJob        = "job"
+)
 
+// StorageBreakdown describes the storage requirements of a directive
+type StorageBreakdown struct {
 	// Lifetime is the duration of the allocation
 	// +kubebuilder:validation:Enum=job;persistent
 	Lifetime string `json:"lifetime"`
+
+	// Reference is an ObjectReference to another resource
+	Reference corev1.ObjectReference `json:"reference,omitempty"`
+
+	// AllocationSets lists the allocations required to fulfill the #DW Directive
+	AllocationSets []StorageAllocationSet `json:"allocationSets,omitempty"`
+}
+
+const (
+	ComputeLocationNetwork  = "network"
+	ComputeLocationPhysical = "physical"
+)
+
+// ComputeLocationConstraint describes a constraints on which compute nodes can be used with
+// a directive based on their location
+type ComputeLocationConstraint struct {
+	// Type is the relationship between the compute nodes and the resource in the Reference
+	// +kubebuilder:validation:Enum=physical;network
+	Type string `json:"type"`
+
+	// Reference is an object reference to a resource that contains the location information
+	Reference corev1.ObjectReference `json:"reference"`
+}
+
+// ComputeConstraints describes the constraints to use when picking compute nodes
+type ComputeConstraints struct {
+	// Location is a list of location constraints
+	Location []ComputeLocationConstraint `json:"location,omitempty"`
+}
+
+// ComputeBreakdown describes the compute requirements of a directive
+type ComputeBreakdown struct {
+	// Constraints to use when picking compute nodes
+	Constraints ComputeConstraints `json:"constraints,omitempty"`
+}
+
+// DirectiveBreakdownSpec defines the storage information WLM needs to select NNF Nodes and request storage from the selected nodes
+type DirectiveBreakdownSpec struct {
+	// Directive is a copy of the #DW for this breakdown
+	Directive string `json:"directive"`
```

Comment:
You removed the dwIndex here. Would that be inferred from the name? I haven't looked at the changes yet to the nnfworkflow controller.
I originally added the index into the directive breakdown to make it easier to identify which breakdown went with which directive in the workflow in case there were repeating directives of the same kind (jobdw, ...).

Reply:
If the think the only time it would be necessary to know which index a directive was would be for a human who was looking at the directive breakdown. In that case they could infer it from the name if they wanted to know.